cpp/include/cuspatial/experimental/detail/derive_trajectories.cuh

```diff
@@ -0,0 +1,149 @@
+/*
+ * Copyright (c) 2022, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "cub/device/device_merge_sort.cuh"
+#include "cub/device/device_radix_sort.cuh"
+#include "cub/device/device_segmented_sort.cuh"
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/device_uvector.hpp>
+#include <rmm/device_vector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/gather.h>
+#include <thrust/iterator/constant_iterator.h>
+#include <thrust/iterator/discard_iterator.h>
+#include <thrust/iterator/zip_iterator.h>
+#include <thrust/reduce.h>
+#include <thrust/sort.h>
+
+#include <cstdint>
+
+namespace cuspatial {
+
+namespace detail {
+
+template <typename IdInputIt,
+          typename PointInputIt,
+          typename TimestampInputIt,
+          typename IdOutputIt,
+          typename PointOutputIt,
+          typename TimestampOutputIt>
+void order_trajectories(IdInputIt ids_first,
+                        IdInputIt ids_last,
+                        PointInputIt points_first,
+                        TimestampInputIt timestamps_first,
+                        IdOutputIt ids_out_first,
+                        PointOutputIt points_out_first,
+                        TimestampOutputIt timestamps_out_first,
+                        rmm::cuda_stream_view stream,
+                        rmm::mr::device_memory_resource* mr)
+{
+  using id_type        = typename std::iterator_traits<IdInputIt>::value_type;
+  using timestamp_type = typename std::iterator_traits<TimestampInputIt>::value_type;
+  using tuple_type     = thrust::tuple<id_type, timestamp_type>;
+
+  auto key_comparator = [] __device__(tuple_type const& lhs, tuple_type const& rhs) {
+    auto lhs_id = thrust::get<0>(lhs);
+    auto rhs_id = thrust::get<0>(rhs);
+    auto lhs_ts = thrust::get<1>(lhs);
+    auto rhs_ts = thrust::get<1>(rhs);
```

Comment:
I was curious the difference between this code as written, and the suggestion here, so I compared them in godbolt:

* current: https://godbolt.org/z/z3G3qrMa6
* with refs: https://godbolt.org/z/EWbjTYYc6

They're quite different! I'm not exactly sure what's going on, but it seems like the second one compiles down to a series of chained branches? Am I reading this right?

```suggestion
    auto& lhs_id = thrust::get<0>(lhs);
    auto& rhs_id = thrust::get<0>(rhs);
    auto& lhs_ts = thrust::get<1>(lhs);
    auto& rhs_ts = thrust::get<1>(rhs);
```

Reply:
I don't see any reason to change this to references. In the case of `float` coordinates, a reference may actually be a bigger value (64-bit pointer vs 32-bit float)?  In the case of `double` coordinates, the reference is still no smaller.

Reply:
Yeah that's the sort of thing I was trying to check in godbolt. Didn't mean to suggest changing it, just using suggestion here to get a nice diff output. I've used refs to `thrust::get` vars in kernels before, but never been sure if it made any difference to performance.

Reply:
I imagine it *could* help in some cases.